numpy/core/src/umath/matmul.c.src

```diff
@@ -0,0 +1,384 @@
+     * op:  contiguous data in c order, m shape
+     */
+    enum CBLAS_ORDER order;
+    int lda;
+
+    if (is1_m == sizeof(@type@)) {
+        order = CblasColMajor;
+        lda = n;
+    }
+    else {
+        /* If not ColMajor, caller should have ensured we are RowMajor */
+        /* will not assert in release mode */
+        assert(is1_m == sizeof(@type@) * m);
+        order = CblasRowMajor;
+        lda = m;
+    }
+    cblas_@prefix@gemv(order, CblasTrans, n, m, @step1@, ip1, lda, ip2,
+                                     is2_n / sizeof(@type@), @step0@, op, 1);
+}
+
+NPY_NO_EXPORT void
+@name@_matmul_matrixmatrix(void *ip1, npy_intp is1_m, npy_intp is1_n,
+                           void *ip2, npy_intp is2_n, npy_intp is2_p,
+                           void *op, npy_intp m, npy_intp n, npy_intp p)
+{
+    /*
+     * matrix matrix multiplication -- Level 3 BLAS
+     */
+    enum CBLAS_ORDER order = CblasRowMajor;
+    enum CBLAS_TRANSPOSE trans1, trans2;
+    int M, N, P, lda, ldb;
+    M = m;
+    N = n;
+    P = p;
+
+    if (is1_m == sizeof(@type@)) {
+        trans1 = CblasTrans;
+        lda = N > 1 ? is1_n / sizeof(@type@) : 1;
+    }
+    else {
+        /* If not ColMajor, caller should have ensured we are RowMajor */
+        /* will not assert in release mode */
+        assert(is1_n == sizeof(@type@));
+        trans1 = CblasNoTrans;
+        lda = N > 1 ? is1_m / sizeof(@type@) : 1;
+    }
+    if (is2_n == sizeof(@type@)) {
+        trans2 = CblasTrans;
+        ldb = N > 1 ? is2_p / sizeof(@type@) : 1;
+    }
+    else {
+        /* If not ColMajor, caller should have ensured we are RowMajor */
+        /* will not assert in release mode */
+        assert(is2_p == sizeof(@type@));
+        trans2 = CblasNoTrans;
+        ldb = P > 1 ? P : 1;
+    }
+    /*
+     * Use syrk if we have a case of a matrix times its transpose.
+     * Otherwise, use gemm for all other cases.
+     */
+    if (
+        (ip1 == ip2) &&
+        (m == p) &&
+        (is1_m == is2_p) &&
+        (is1_n == is2_n) &&
+        (trans1 != trans2)
+    ) {
+        npy_intp i,j;
+        if (trans1 == CblasNoTrans) {
+            cblas_@prefix@syrk(order, CblasUpper, trans1, P, N, @step1@,
+                               ip1, lda, @step0@, op, P);
+        }
+        else {
+            cblas_@prefix@syrk(order, CblasUpper, trans1, P, N, @step1@,
+                               ip1, ldb, @step0@, op, P);
+        }
+        /* Copy the triangle */
+        for (i = 0; i < P; i++) {
+            for (j = i + 1; j < P; j++) {
+                ((@type@*)op)[j * P + i] = ((@type@*)op)[i * P + j];
+            }
+        }
+
+    }
+    else {
+        cblas_@prefix@gemm(order, trans1, trans2, M, P, N, @step1@, ip1, lda,
+                           ip2, ldb, @step0@, op, P);
+    }
+}
+
+/**end repeat**/
+#endif
+
+/*
+ * matmul loops
+ * signature is (m?,n),(n,p?)->(m?,p?)
+ */
+
+/**begin repeat
+ * #TYPE = FLOAT, DOUBLE, HALF#
```

Comment:
What's special about longdouble here that means it's ignored?

Reply:
It is handled in the next repeat set. This set uses a double intermediate value.

Reply:
That's what I'm asking - why special case a double intermediate value here rather than using a value of the same type? A comment explaining the intent would be valuable.

Reply:
documenting why the choice was made, perhaps a different choice could also be made

Reply:
I think it might be neater to combine the repeats, and use macros to toggle this double-intermediate behavior on and off. I'm not a massive fan of repeating the entire inner loop multiple times manually, when our template system is just about capable of handling it.

Reply:
OK, merging the two loops